### PR TITLE
storage: Add API V2 check for RawKV and TxnKV requests (txn part)

### PR DESCRIPTION
**Pull request by pingyu**
### What problem does this PR solve?

Issue Number: #10974 

Problem Summary:

Add API V2 check for RawKV and TxnKV requests.

### What is changed and how it works?

Proposal: [TiKV API V2](https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md)

What's Changed:

Add checker logics for the following API:

- prewrite
- pessimistic lock

### Related changes

No.

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Breaking backward compatibility
Breaking backward compatibility: See rfc [TiKV API V2](https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md) for detail.

### Release note 

```release-note
None.
```

### CODE REVIEW

**Comment by pingyu** [comment]
@andylokandy 
Merged changes of #11228, and modified api checking for TTL storage.
PTAL~

**Comment by andylokandy** [comment]
Maybe we can check txn keys here: https://github.com/tikv/tikv/blob/8eb9cb81909df74e839504559d3325736cd28aea/src/storage/mod.rs#L1160-L1181

The advantage is that check_api_version does not leak outside of storage.

**Comment by pingyu** [comment]
> Maybe we can check txn keys here:
> 
> https://github.com/tikv/tikv/blob/8eb9cb81909df74e839504559d3325736cd28aea/src/storage/mod.rs#L1160-L1181
> 
> The advantage is that check_api_version does not leak outside of storage.

Done. PTAL, thanks~ @andylokandy

File: src/storage/mod.rs [hunk 1]
```diff
@@ -1171,25 +1156,37 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
 
         match &cmd {
             Command::Prewrite(Prewrite { mutations, .. }) => {
-                check_key_size!(
-                    mutations.iter().map(|m| m.key().as_encoded()),
-                    self.max_key_size,
-                    callback
-                );
+                let keys: Vec<&Vec<u8>> = mutations.iter().map(|m| m.key().as_encoded()).collect();
```
Comment (andylokandy) [inline src/storage/mod.rs:1159]: Is the collect() avoidable?

**Review by andylokandy — APPROVED** [review]
rest LGTM

**Comment by andylokandy** [comment]
@iosmanthus @zhongzc  PTAL

File: tests/integrations/server/kv_service.rs [hunk 2]
```diff
@@ -1744,3 +1746,207 @@ fn test_get_lock_wait_info_api() {
...
+        ),
+        // config api_version = V2.
+        // backward compatible for TiDB request, and TiDB request only.
+        (ApiVersion::V2, ApiVersion::V1, TIDB_KEY_CASE, None),
+        (
+            ApiVersion::V2,
+            ApiVersion::V1,
+            TXN_KEY_CASE,
+            Some("InvalidKeyPrefix"),
+        ),
```
Comment (zhongzc) [inline tests/integrations/server/kv_service.rs:1782]: How about checking RAW_KEY_CASE also?

File: tests/integrations/server/kv_service.rs [hunk 3]
```diff
@@ -1744,3 +1746,207 @@ fn test_get_lock_wait_info_api() {
...
     handle.join().unwrap();
 }
+
+#[test]
+fn test_api_version() {
+    const TIDB_KEY_CASE: &[u8] = b"t_a";
+    const TXN_KEY_CASE: &[u8] = &[key_prefix::TXN_KEY_PREFIX, 0, b'a'];
+    const RAW_KEY_CASE: &[u8] = &[key_prefix::RAW_KEY_PREFIX, 0, b'a'];
+
+    let test_data = vec![
```
Comment (zhongzc) [inline tests/integrations/server/kv_service.rs:1756]: The APIv2 spec still confuse me. Can you explain the relation between APIv2 and RAW?

And I think it's better to add some comments about the spec here.

Comment (andylokandy) [reply tests/integrations/server/kv_service.rs:1756]: Is this spec sufficient?
https://github.com/pingcap/kvproto/blob/d0692eebe1ed1827811b24a377a45b3bed58f15b/proto/kvrpcpb.proto#L715-L762

File: tests/integrations/server/kv_service.rs [hunk 4]
```diff
@@ -1744,3 +1746,207 @@ fn test_get_lock_wait_info_api() {
...
+        // config api_version = V1|V1ttl, for backward compatible.
+        (ApiVersion::V1, ApiVersion::V1, TIDB_KEY_CASE, None),
+        (ApiVersion::V1, ApiVersion::V1, TXN_KEY_CASE, None),
```
Comment (zhongzc) [inline tests/integrations/server/kv_service.rs:1759]:
```suggestion
        (ApiVersion::V1, ApiVersion::V1, TIDB_KEY_CASE, None),
        (ApiVersion::V1, ApiVersion::V1, TXN_KEY_CASE, None),
        (ApiVersion::V1, ApiVersion::V1, RAW_KEY_CASE, None),
```

File: tests/integrations/server/kv_service.rs [hunk 5]
```diff
@@ -1744,3 +1746,207 @@ fn test_get_lock_wait_info_api() {
...
+        // storage api_version = V1ttl, allow RawKV request only.
+        (
+            ApiVersion::V1ttl,
```
Comment (zhongzc) [inline tests/integrations/server/kv_service.rs:1762]: We can set the storage API version as `ApiVersion::V1ttl`? @andylokandy

Comment (pingyu) [reply tests/integrations/server/kv_service.rs:1762]: Yes. Set `enable-ttl: true` and `api-version: 1` (or leave `api-version` as default)

*(codes [here](https://github.com/tikv/tikv/blob/1e1f32ab744cafd0145cd93e48249025455f6ae7/src/storage/config.rs#L121-L146))*
```
    pub fn api_version(&self) -> ApiVersion {
        match self.api_version {
            1 if self.enable_ttl => ApiVersion::V1ttl,
            1 => ApiVersion::V1,
            2 => ApiVersion::V2,
            _ => unreachable!(),
        }
    }

    pub fn set_api_version(&mut self, api_version: ApiVersion) {
        match api_version {
            ApiVersion::V1 => {
                self.api_version = 1;
                self.enable_ttl = false;
            }
            ApiVersion::V1ttl => {
                self.api_version = 1;
                self.enable_ttl = true;
            }
            ApiVersion::V2 => {
                self.api_version = 2;
                self.enable_ttl = false;
            }
        }
    }
```

**Comment by pingyu** [comment]
Fixed, and improved other similar test cases.
PTAL, thanks~ @zhongzc

File: tests/integrations/server/kv_service.rs (hunk 5 above)
Comment (andylokandy) [reply tests/integrations/server/kv_service.rs:1767]: Oh, `ApiVersion::V2` should set  `self.enable_ttl = true;`! Would you like to correct it?

**Review by zhongzc — APPROVED** [review]

**Pull request merged**